Next version

n0cte opened the pull request.

n0cte committed:
add another implementation

kushkamisha committed:
Fix linter errors

+ removed invalid `recover` test for ProofOfAgreement as there is no signature for ProofOfAgreement

kushkamisha committed:
Fix domain in proofs metadata

kushkamisha committed:
Fix Signer field

kushkamisha committed:
Add Shrinked structs

kushkamisha committed:
Fix timestamp type

kushkamisha committed:
Fix ProoofOfAuthority.fileCID -> agreementFileCID

kushkamisha committed:
Fix ProofOfSignature.fileCid -> agreementFileProofCID

kushkamisha committed:
Fix ProofOfAgreement structure

kushkamisha committed:
Add StringUtils library

kushkamisha committed:
Store ProofOfAuthority func + test

kushkamisha committed:
Bring old way of storing proofs

kushkamisha committed:
Store function impl and more

kushkamisha committed:
Add ProofsBase contract

kushkamisha committed:
Add full validation for Proof-of-Signature and Proof-of-Agreement

kushkamisha committed:
Small improvements

n0cte committed:
do restructurization

n0cte committed:
add logic reimplementation

n0cte committed:
add correct address comparison

n0cte committed:
add "after hook" after contract compilation

n0cte committed:
add hash readers for mock contract

n0cte committed:
Merge branch 'main' into reimpl

kushkamisha committed:
Layout ordering
* Move all functions, structs, events inside the contracts
* Remove DAOSignBaseApp interface as it's not used anywhere except for DAOSignApp
* Run linter

kushkamisha committed:
Some more general fixes
* Add/change state visibility
* Explicit imports

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+    mapping(string => mapping(address => uint256)) poauSignersIdx;
+
+    constructor() {
+        domain.name = 'daosign';
+        domain.version = '0.1.0';
+        DOMAIN_HASH = hash(domain);
```
domain.version may be updated in the future. We'll need some setters for that

n0cte replied:
"setter" it is state change, not upgrade. i think, "Proxy Upgrade Pattern" is better choise

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+        initEIP712Types();
+    }
+
+    function memcmp(bytes memory a, bytes memory b) internal pure returns (bool) {
+        return (a.length == b.length) && (keccak256(a) == keccak256(b));
+    }
+
+    function strcmp(string memory a, string memory b) internal pure returns (bool) {
+        return memcmp(bytes(a), bytes(b));
+    }
```
These functions are not related to the DAOSignApp scope. I would move them to library or abstract contract

n0cte replied:
it is pointless to create a library for one function

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+            });
+    }
+
+    function getProofOfAgreement(
+        string memory cid
+    ) external view returns (SignedProofOfAgreementMsg memory) {
+        SignedProofOfAgreement memory data = poags[cid];
+        return
+            SignedProofOfAgreementMsg({
+                message: toEIP712Message(data.message),
+                signature: data.signature
+            });
+    }
+}
```
There is no signature of Proof-of-Agreement. So you may update the returning value

n0cte replied:
This is a generalized structure, needed for uniformity of answers

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+
+    function validate(SignedProofOfAuthority memory data) internal pure returns (bool) {
+        require(bytes(data.proofCID).length == IPFS_CID_LENGHT, 'Invalid proof CID');
+        require(strcmp(data.message.app, 'daosign'), 'Invalid struct name');
+        require(strcmp(data.message.name, 'Proof-of-Authority'), 'Invalid struct name');
```
Here you have the same error message "Invalid struct name" used twice. It is also used in functions later. It's better to have unique error messages

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+        return memcmp(bytes(a), bytes(b));
+    }
+
+    function validate(SignedProofOfAuthority memory data) internal pure returns (bool) {
```
Here and in the following functions, I would validate data.message.timestamp

n0cte replied:
@ramilexe

ramilexe replied:
@kushkamisha what kind of validation do want to add for timestamp?

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+        return true;
+    }
+
+    function validate(SignedProofOfAgreement memory data) internal view returns (bool) {
```
data.message.timestamp in Proof-of-Agreement should be equal to the last Proof-of-Signature's timestamp. You may want to validate that

n0cte replied:
@ramilexe This is not a true statement. The creation time of the "Proof-of-Agreement" data will never be equal to the time of the last signature

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+                data.message.signatureCIDs.length,
+            'Invalid sign proofs'
+        );
+        for (uint i = 0; i < data.message.signatureCIDs.length; i++) {
```
The order of signers in Proof-of-Authority and in Proof-of-Agreement may not be the same, so it's better not to rely on that

ramilexe replied:
good catch, we need to sort it first

n0cte replied:
it uses index checking, please take a closer look at the algorithm

ramilexe replied:
@n0cte @kushkamisha can we check in tests?

n0cte replied:
yes, we can add this point into unit tests

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+    function storeProofOfAuthority(SignedProofOfAuthority memory data) external {
+        require(recover(data.message, data.signature) == data.message.from, 'Invalid signature');
+        require(validate(data), 'Invalid message');
+        poaus[data.proofCID].message.name = data.message.name;
+        poaus[data.proofCID].message.from = data.message.from;
+        poaus[data.proofCID].message.agreementCID = data.message.agreementCID;
+        poaus[data.proofCID].message.app = data.message.app;
+        poaus[data.proofCID].message.timestamp = data.message.timestamp;
+        poaus[data.proofCID].message.metadata = data.message.metadata;
+        for (uint i = 0; i < data.message.signers.length; i++) {
+            poauSignersIdx[data.proofCID][data.message.signers[i].addr] = i;
+            poaus[data.proofCID].message.signers.push(data.message.signers[i]);
+        }
+        poaus[data.proofCID].signature = data.signature;
+        poaus[data.proofCID].proofCID = data.proofCID;
```
Instead of all these lines of code why you're not using `abi.encode(...)` to do `poaus[data.proofCID] = abi.encode(...)` as you've initially suggested?

n0cte replied:
it’s cheaper and you don’t have to constantly decode when reading

kushkamisha commented on contracts/DAOSignApp.sol:
```diff
@@ -0,0 +1,204 @@
+        emit NewProofOfSignature(data);
+    }
+
+    function storeProofOfAgreement(SignedProofOfAgreement memory data) external {
```
You may want to rename SignedProofOfAgreement into something like ProofOfAgreement as Proof-of-Agreement doesn't contain any signature as it's not signed

n0cte replied:
1. this is the unification and generalization of messages
2. since we want to forward messages, this is additional protection
3. in the future, this will make it possible to assign closing messages to different applications

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+struct EIP712Domain {
+    string name;
+    string version;
+    uint256 chainId;
+    address verifyingContract;
```
We probably don't need chainId and verifyingContract as we may be storing the same proof on Ethereum, Polkadot, and SUI where the chain Ids are different + we don't have a verifying contract either way. We may just leave these variables empty but, as I see it, this will just create unnecessary extra complexity.

n0cte replied:
in the future we are required to implement "verifyingContract". The sui network will support different wallets, for this we can use "chainId"

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+struct ProofOfSignature {
+    string name;
+    address signer;
+    string agreementCID;
```
It shouldn't be agreementCID, instead it should be CID of Proof-of-Authority. It has to be updated here + some logic in DAOSignApp may need to be adjusted as well

ramilexe replied:
Regarding our [technical design](https://consideritdone.atlassian.net/wiki/spaces/DAOS/pages/2561933440/Proofs+Technical+Design) it seems we use `agreementCID` there

ramilexe replied:
![image](https://github.com/DAOsign/ethereum-smart-contracts/assets/1008882/6722a9af-c521-46a4-9d25-e9dd44b00d23)
I would rename this field to `agreementFileProofCID`

kushkamisha replied:
According to this document, this is what should be in the Proof-of-Signature message:
```
"message": {
    "name": "Proof-of-Signature",
    "signer": "<signer's address>",
    "agreementFileProofCID": "<Agreement File Proof CID>",
    "app": "daosign",
    "timestamp": <timestamp in seconds>,
    "metadata": "{}"
  }
  ```
  
And Agreement File Proof CID is IPFS CID of the signed Agreement File Proof (a.k.a Proof-of-Authority)

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+);
+
+struct ProofOfAgreement {
+    string agreementCID;
```
Same here. You should have proofOfAuthorityCID instead of agreementCID

ramilexe replied:
It seems we have names misleading in tech design and smart contracts

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+    EIP712PropertyType[7] ProofOfAuthority;
+}
+
+struct EIP712ProofOfAuthority {
+    EIP712ProofOfAuthorityTypes types;
+    EIP712Domain domain;
+    string primaryType;
+    ProofOfAuthority message;
+}
```
Have you tested that this is displayed properly in MetaMask pop up? For example, I'm not sure that changing "name" - "type" pair to "name" - "kind" will be formatted in MetaMask properly

n0cte replied:
check added in tests

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+    ProofOfSignature message;
+}
+
+struct EIP712ProofOfAgreementTypes {
+    EIP712PropertyType[4] EIP712Domain;
+    EIP712PropertyType[5] ProofOfAgreement;
+}
+
+struct EIP712ProofOfAgreement {
+    EIP712ProofOfAgreementTypes types;
+    EIP712Domain domain;
+    string primaryType;
+    ProofOfAgreement message;
+}
```
Proof-of-Agreement doesn't have EIP712 types as it's not signed. It should look like this:
```
{
  "agreementFileProofCID": "<Agreement File Proof CID>",
  "agreementSignProofs": [
      { "proofCID": "<Agreement Sign Proof CID>" },
      { "proofCID": "<Agreement Sign Proof CID>" },
      { "proofCID": "<Agreement Sign Proof CID>" }
  ],
  "timestamp": <timestamp in seconds; this this the last signature's timestamp>
}
```

n0cte replied:
that message should be created via our backend

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+            kind: 'address'
+        });
+        proofOfSignatureDoc.types.ProofOfSignature[2] = EIP712PropertyType({
+            name: 'agreementCID',
```
agreementCID should contain Proof-of-Authority CID so it should be named in other way

kushkamisha commented on contracts/DAOSignEIP712.sol:
```diff
@@ -0,0 +1,426 @@
+        return recover(digest, signature);
+    }
+
+    function recover(
+        ProofOfAgreement memory data,
+        bytes memory signature
+    ) internal view returns (address) {
+        bytes32 packetHash = hash(data);
+        bytes32 digest = keccak256(abi.encodePacked('\x19\x01', DOMAIN_HASH, packetHash));
+        return recover(digest, signature);
+    }
```
No need to recover Proof-of-Agreement as there is no signature

kushkamisha committed:
Made DAOSignEIP712 abstract

kushkamisha committed:
Remove unused dependencies

kushkamisha committed:
ZEROADDR -> ethers.ZeroAddress

kushkamisha commented on tasks/compile.ts:
```diff
@@ -0,0 +1,21 @@
+import { existsSync, readFileSync, writeFileSync } from 'fs';
+import { extendEnvironment } from 'hardhat/config';
+
+extendEnvironment(function (hre) {
+  const fnames = [
+    './artifacts/contracts/DAOSignApp.sol/DAOSignApp.json',
+    './artifacts/contracts/DAOSignApp.sol/DAOSignBaseApp.json',
+    './artifacts/contracts/mocks/MockDAOSignEIP712.sol/MockDAOSignEIP712.json',
+  ];
+  const action = hre.tasks['typechain:generate-types'].action;
+  hre.tasks['typechain:generate-types'].setAction(async function (...params) {
+    for (const fname of fnames) {
+      if (!existsSync(fname)) {
+        throw new Error(`file "${fname}" doesn't exist`);
+      }
+      const fdata = readFileSync(fname, { encoding: 'utf-8' });
+      writeFileSync(fname, fdata.replaceAll('kind', 'type'));
+    }
+    return action(...params);
+  });
+});
```
I would avoid any logic like this as now you have to always edit this file manually if you start testing new contracts which may not be intuitive

n0cte replied:
please suggest a different logic for renaming fields. for example: we can use file prefixes of suffixer for field's name modification

kushkamisha committed:
validate Proof-of-Authority tests

kushkamisha committed:
validate Proof-of-Signature tests

kushkamisha committed:
validate Proof-of-Agreement tests

n0cte committed:
fix error msg

kushkamisha committed:
Add explicit EIP712Domain initialization

kushkamisha committed:
Undo some changes from the last commit

n0cte committed:
add deployment task

kushkamisha committed:
Fixes

- created IDAOSignApp inerface and moved it to `interfaces` folder
- moved structs outside of smart contracts

n0cte committed:
Merge pull request #3 from DAOsign/review

Review of reimplementation
